File: modules/covariate-mic/android/src/main/java/expo/modules/covariatemic/CovariateMicModule.kt
package expo.modules.covariatemic

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Stub — Android mic-level capture (e.g. via AudioRecord, computing RMS
 * in-memory the same way the iOS AVAudioEngine tap does) has not been
 * ported yet. Reports unavailable rather than crashing on a missing native
 * method. Flagging as an open item: unclear whether the cross-device
 * reliability study needs Android parity here or is iPhone-only — worth
 * confirming before this ships, since it's a real scope decision, not
 * just an implementation detail.
 */
class CovariateMicModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CovariateMic")

    Events("onSample")

    AsyncFunction("requestPermissionsAsync") {
      mapOf("granted" to false, "status" to "unavailable")
    }

    AsyncFunction("start") {
      throw Exception("Mic-level sensing is not implemented on Android in this module yet.")
    }

    Function("stop") {
      // no-op
    }
  }
}

File: modules/covariate-mic/src/CovariateMicModule.ts
import { NativeModule, requireNativeModule } from 'expo';
import type { CovariateMicModuleEvents } from './CovariateMic.types';

declare class CovariateMicModule extends NativeModule<CovariateMicModuleEvents> {
  /** Prompts for microphone permission if not yet determined (iOS only; always unavailable on Android). */
  requestPermissionsAsync(): Promise<{ granted: boolean; status: string }>;
  /** Starts the AVAudioEngine tap. Samples arrive via the 'onSample' event, not a return value. Never writes an audio file. */
  start(): Promise<void>;
  stop(): void;
}

export default requireNativeModule<CovariateMicModule>('CovariateMic');
